Add documentation for EBT replication

mycognosist opened the pull request:
This PR builds upon the excellent work done by @boreq on the Planetary [Epidemic Broadcast Trees documentation](http://dev.planetary.social/replication/ebt.html); I would not have been able to successfully implement EBT without them. I have also drawn on the documentation in the [epidemic-broadcast-trees](https://github.com/ssbc/epidemic-broadcast-trees) `README`.

My work is still in draft form but I am inviting feedback at this stage. A few questions to guide the conversation:

1. Is the Protocol Guide the appropriate place for these docs? Or should I rather submit an update to the Planetary docs and also include them in the [Scuttlebutt Treasure Map for Developers](https://dev.scuttlebutt.nz)?
2. Is the content correct?
3. Is the content clear?
4. Is the content appropriately positioned in the guide (ie. immediately after `createHistoryStream` in the `Feeds` section)?
5. Does the writing style blend seamlessly with the rest of the guide?
6. How can the encoding / decoding table be improved?

I have tried where possible to mimic the style employed in the rest of the guide.

The colours of the encoding / decoding table have been chosen in relation to the other colours used in the guide, ie. blue for requests, orange for responses etc.

Input is welcome!

mycognosist committed:
add basics of ebt replication

mycognosist committed:
add styling for vector clock value table

mycognosist committed:
add sections on session termination and request skipping

mycognosist committed:
add section on fallback mechanisms

mycognosist committed:
further copy-editing

mycognosist committed:
make encoded values bold in vector clock table

mycognosist committed:
revert to nth-child styling for vector clock table

ahdinosaur commented:
btw, for anyone else, here's a link to preview: https://htmlpreview.github.io/?https://raw.githubusercontent.com/ssbc/scuttlebutt-protocol-guide/7ec9c41f7a504af92b4f441c4e8c1d3e7cb589d5/index.html#ebt-replication

ahdinosaur commented:
looks great @mycognosist ! i certainly feel like i understand EBT better than i did before.

i'm not sure i have much useful feedback, but can at least try to respond to your questions.

> 1. Is the Protocol Guide the appropriate place for these docs? Or should I rather submit an update to the Planetary docs and also include them in the [Scuttlebutt Treasure Map for Developers](https://dev.scuttlebutt.nz)?

personally, yes i think the Protocol Guide _is_ the appropriate place for these docs.

> 2. Is the content correct?

i'm no EBT expert so :shrug: 

> 3. Is the content clear?

i feel so yes.

> 4. Is the content appropriately positioned in the guide (ie. immediately after `createHistoryStream` in the `Feeds` section)?

that position makes the most sense to me.

> 5. Does the writing style blend seamlessly with the rest of the guide?

i feel so yes. i think your language is as clear and straightforward as the rest of the guide.

> 6. How can the encoding / decoding table be improved?

i think what's missing is a larger sequence number. in your example above you have 450, 12, 1, and -1. i'd like to see what those decode to.

staltz commented:
> 1. Is the Protocol Guide the appropriate place for these docs?

Yes, absolutely

> 2. Is the content correct?

I can't say, since other than Dominic's brain we haven't a *standard* against which we can compare. I think it's acceptable to merge this and if we learn of inaccuracies (the protocol guide is a description of SSB as is, not a specification of how SSB should be) then we can update the guide in the future.

> Is the content clear?

Let me put it like this: it's like EBT was dark basement and suddenly someone finally turned on the lights.

> Is the content appropriately positioned in the guide (ie. immediately after `createHistoryStream` in the `Feeds` section)?

Yes, totally

> Does the writing style blend seamlessly with the rest of the guide?

There's heavy usage of the word "should" so a keen reader might notice a difference in writing style, but not a problem by any means. Perhaps it's easy to tweak the text to use less "should"s, if you feel like blending the writing style further.

> How can the encoding / decoding table be improved?

Pretty clear to me, I don't know how to improve it.

cblgh commented:
Finally: EBT revealed!

> Is the Protocol Guide the appropriate place for these docs? Or should I rather submit an update to the Planetary docs and also include them in the [Scuttlebutt Treasure Map for Developers](https://dev.scuttlebutt.nz)?

Yup!

>Is the content correct?

As far as I can tell :)

> Is the content clear?

Yup! Very clear, with succinct descriptions that remain quickly readable 

> Is the content appropriately positioned in the guide (ie. immediately after `createHistoryStream` in the `Feeds` section)?

Yup! After createHistoryStream seems like a logical place

> Does the writing style blend seamlessly with the rest of the guide?
As far as I can tell :)

> * How can the encoding / decoding table be improved?

I think it's good?

---

A morsel of feedback:

In Section "Request Skipping", sidebar note: "Approaches to tracking EBT session state may be gleaned from the JS and Go implementations."

I think this note could link to the relevant (pinned?) lines / files in the JS / Go implementations, as is done for other notes in the guide? Would make it easier for potential implementers to track down the code :)

quickdudley reviewed (approved):
A client implementing the protocol as described here would sync successfully with Habitat.

mycognosist committed:
add decode examples for 12 and 450

mycognosist committed:
reduce usage of the word 'should'

mycognosist commented:
Thanks for all the terrific feedback! 🖤 

> i think what's missing is a larger sequence number. in your example above you have 450, 12, 1, and -1. i'd like to see what those decode to.

@ahdinosaur : Great idea! I've added examples for `12` and `450`.

> There's heavy usage of the word "should" so a keen reader might notice a difference in writing style

@staltz : Good catch. I slipped into RFC language there. I've replaced most of the instances of "should" and I think it blends slightly better now.

> I think this note could link to the relevant (pinned?) lines / files in the JS / Go implementations, as is done for other notes in the guide? Would make it easier for potential implementers to track down the code :)

@cblgh : I like the idea but it's a little tricky because state management is quite spread out in both implementations (involving multiple modules in Go and several functions / methods in JS). I'm hoping that my links to both implementations at the start of the EBT section will at least narrow the search radius ^_^

> A client implementing the protocol as described here would sync successfully with Habitat.

@quickdudley : That's very reassuring. Thanks for taking the time to read and comment.

-----

I'll give another 24 hours or so for comment and then I think it's safe to go ahead with a merge.

mycognosist commented:
I see no further comments. Going ahead with the merge. Thanks folx.
